pkg/networking/gwconnection/connection.go

```diff
@@ -4,49 +4,216 @@
 package gwconnection
 
 import (
+	"context"
 	"fmt"
+	"net"
+	"net/netip"
 	"time"
 
 	"go.uber.org/zap"
 
-	ping "github.com/prometheus-community/pro-bing"
+	types100 "github.com/containernetworking/cni/pkg/types/100"
+	"github.com/containernetworking/plugins/pkg/ns"
+	"github.com/mdlayher/arp"
+	"github.com/mdlayher/ethernet"
+	"github.com/mdlayher/ndp"
 )
 
-type Pinger struct {
-	logger *zap.Logger
-	pinger *ping.Pinger
+type DetectGateway struct {
+	retries                    int
+	iface                      string
+	interval                   time.Duration
+	timeout                    time.Duration
+	v4Addr, v6Addr, V4Gw, V6Gw net.IP
+	logger                     *zap.Logger
+	hostNs, podNs              ns.NetNS
 }
 
-func NewPinger(count int, interval, timeout, gw string, logger *zap.Logger) (*Pinger, error) {
-	pinger := ping.New(gw)
-	pinger.Count = count
+func New(retries int, interval, timeout, iface string, logger *zap.Logger) (*DetectGateway, error) {
+	var err error
+	dg := &DetectGateway{
+		retries: retries,
+		iface:   iface,
+	}
 
-	intervalDuration, err := time.ParseDuration(interval)
+	dg.interval, err = time.ParseDuration(interval)
 	if err != nil {
 		return nil, err
 	}
-	pinger.Interval = intervalDuration
 
-	timeoutDuration, err := time.ParseDuration(timeout)
+	dg.timeout, err = time.ParseDuration(timeout)
 	if err != nil {
 		return nil, err
 	}
-	pinger.Timeout = timeoutDuration
-	pinger.SetPrivileged(true)
 
-	return &Pinger{logger, pinger}, nil
+	return dg, nil
+}
+
+func (dg *DetectGateway) ParseAddrFromPreresult(ipconfigs []*types100.IPConfig) {
+	for _, ipconfig := range ipconfigs {
+		if ipconfig.Address.IP.To4() != nil {
+			dg.v4Addr = ipconfig.Address.IP
+		} else {
+			dg.v6Addr = ipconfig.Address.IP
+		}
+	}
 }
 
-func (p *Pinger) DetectGateway() error {
-	if err := p.pinger.Run(); err != nil {
-		return fmt.Errorf("failed to run DetectGateway: %v", err)
+// PingOverIface sends an arp ping over interface 'iface' to 'dstIP'
+func (dg *DetectGateway) ArpingOverIface() error {
+	ifi, err := net.InterfaceByName(dg.iface)
+	if err != nil {
+		return err
+	}
+
+	client, err := arp.Dial(ifi)
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	gwNetIP := netip.MustParseAddr(string(dg.V4Gw.String()))
+	// we send a gratuitous arp to checking if ip is conflict
+	// we use dad mode(duplicate address detection mode), so
+	// we set source ip to 0.0.0.0
+	packet, err := arp.NewPacket(arp.OperationRequest, ifi.HardwareAddr, netip.MustParseAddr("0.0.0.0"), ethernet.Broadcast, gwNetIP)
```

Comment:
arp probe is forbidden by some brands of gatway
it should use pod ip for source IP

Reply:
done